[aes/pre_syn] Add experimental Yosys synthesis flow

vogelpi opened the pull request:
This PR adds an experimental Yosys synthesis flow for the AES module under `hw/ip/aes/pre_syn` (similar to the non-sign-off design verification infrastructure we put into `pre_dv`). The idea of this open flow is to generate ASIC area estimates as well as to procude netlists that can further be used for developing and verifying SCA and FI countermeasures, for example by means of formal security verification using tools like REBECCA/Coco-Alma.

The flow itself is based on the experimental Yosys synthesis flow developed for Ibex by @GregAC . I made a couple of minor adjustments to easily change the top module (for formal security verification it is helpful to start verifying small building blocks before going to the S-Box or even the full cipher core). For sure, things could be done in a better and more modular way but I thought it's probably best to get this in sooner than later. We can improve/generalize this as we also start to adopt it for other IPs.

towoe reviewed (commented):
As this reuses code from the Ibex syn flow I did not check everything, but in general this looks good, thanks Pirmin.
And it works on my side :)

towoe reviewed (approved):
As this reuses code from the Ibex syn flow I did not check everything, but in general this looks good, thanks Pirmin.
And it works on my side :)

rswarbrick commented on hw/ip/aes/pre_syn/syn_setup.example.sh:
```diff
@@ -0,0 +1,43 @@
+
+# Setup cell library path.
+# Uncomment the lines below and set the path to an appropriate .lib file
+# export LR_SYNTH_CELL_LIBRARY_PATH=/path/to/NangateOpenCellLibrary_typical.lib
+# export LR_SYNTH_CELL_LIBRARY_NAME=nangate
```
For extra friendliness, it might be good to follow this block with a check like `if [ -z "$LR_SYNTH_CELL_LIBRARY_PATH" ]; echo >&2 'You forgot to set LR_SYNTH_CELL_LIBRARY_PATH!'; exit 1; fi` and similar for the name?

vogelpi replied:
Good idea! I've added this for both these variables.

rswarbrick commented on hw/ip/aes/pre_syn/syn_setup.example.sh:
```diff
@@ -0,0 +1,43 @@
+# For timing runs, hierarchy flattening is needed.
+if [[ $LR_SYNTH_TIMING_RUN == 1 ]] && [[ $LR_SYNTH_FLATTEN == 0 ]] ;
+then
+  echo "Flipping LR_SYNTH_FLATTEN to 1"
+  export LR_SYNTH_FLATTEN=1
```
I don't like this: If TIMING=1 and FLATTEN=0 is not allowed, I think it's probably clearer just to exit with an error message.

vogelpi replied:
Sounds okay to me. Did change the script accordingly.

rswarbrick commented on hw/ip/aes/pre_syn/syn_yosys.sh:
Nit: OpenTitan.

Not for this PR, but surely we could do this by running fusesoc first? Then we'd get all the sources wrapped up nicely. That would also get rid of the sed hacks for technology mapping, I think?

vogelpi replied:
That would be the ideal solution. However, for that we would also need to solve the `.*`-type port assignments in the abstract primitives. This is not supported by Yosys IIRC.

rswarbrick replied:
Have we tried that recently? I'm pretty certain that I've seen the code to support it in the Yosys codebase.

vogelpi replied:
I stumbled across this, but I don't remember if I was using an older Yosys version at that point. I did find a couple of sv2v bugs and at some point decided it was best to also upgrade Yosys. So it could very well be that this works now.

rswarbrick reviewed (approved):
A couple of nits about error handling, but this looks good to me.

rswarbrick commented:
Incidentally, the CI failure is not caused by this patch: if you do a force-push, everything should end up green.

vogelpi commented on hw/ip/aes/pre_syn/aes_lr_synth_conf.tcl:
```diff
@@ -0,0 +1,42 @@
+# clock period in ps, this gives a 250 MHz clock.  using the nangate45 library
+# Ibex can happily meet this on all paths with the lr_synth_abc_clk_uprate
+# setting below. With a lower uprate timing may not be met.
+#set lr_synth_clk_period 4000.0
```
Yep, that's done now.

vogelpi commented on hw/ip/aes/pre_syn/aes_lr_synth_conf.tcl:
```diff
@@ -0,0 +1,42 @@
+# the synth flow. ABC maps the design to the standard cell library and
+# optimises paths for timing, better results are obtained by giving it a faster
+# clock period so it optimises more.
+#set lr_synth_abc_clk_uprate 2000.0
```
Done that, too.

vogelpi commented on hw/ip/aes/pre_syn/syn_yosys.sh:
```diff
@@ -0,0 +1,152 @@
+mkdir -p "$LR_SYNTH_OUT_DIR/reports/timing"
+
+rm -f latest
+ln -s "$LR_SYNTH_OUT_DIR" latest
```
I like the idea. Adapted the script accordingly.

vogelpi commented:
Thanks @towoe and @rswarbrick for your reviews! I've implemented the suggested changes and changed indentation to 4 spaces in the bash scripts (wasn't consistent before). I will merge this once CI passes.
